### DISCUSSION TITLE
Multiplexing?

### DISCUSSION THREAD

**Discussion by mwalkerr** [Textualize/textual#290] [Q&A]
It'd be awesome to embed other terminal applications within an app built with Textual. For example, having a window for vim for quick edits or having the React Native metro server run. Is this something that would be possible with Textual? Are there other libraries I should look to using in conjunction with it?

**Comment by willmcgugan** [comment]
It will be possibly eventually, but it is far from trivial. You would essentially have to process a decent subset of ansi codes. Rich as an AnsiDecoder, but it is not comprehensive.

**Reply by mwalkerr** [reply]
@willmcgugan Do you mean possible at all or possible to do well? Isn't this something that existing terminal multiplexers already handle? I'm probably being overly naive here from not knowing more about terminals, but I would think it's at least possible to have a clunky approach work. My understanding is that the ultimate result of a terminal handling output + ANSI codes from some program (ignoring side-effects like the bell) at a given point in time is to have an nxm grid of text +/- a cursor. If that's the case, I would think that some program like tmux would offer an API for getting the contents of that grid at a given point in time. So in other words, I don't understand yet why something like this wouldn't be possible:
1. Create a Textual app in a terminal that's 160 x 40.
2. Create a widget that's 80x40 and docked to the right
3. Create a tmux session with window size 80x40 and run something like `brew install`
4. Create a thread that gets the current view of that tmux session and writes it to the widget

Granted, even if that does work, it would certainly be less efficient to constantly sync the entire contents and I imagine there's the possibility of catching partially rendered output. I would think a full and clean solution would have to interpret ANSI codes and map them to the appropriate operations within rich/textual, but is it possible that something like described above would work?

**Reply by USLTD** [reply]
Terminal multiplexers use pseudo-terminal (**pty**) to emulated terminal. Meaning that every terminal multiplexer is a terminal emulator so if you want to build multiplexer you need to first understand what is terminal, terminal emulator, pseudo-terminal and ANSI sequences (this last one is optional since it isn't necessary but recommended)
